### ISSUE TITLE
Hosting design inconsistencies and lifetime confusion

### ISSUE THREAD

**Issue by maelgrove** [aspnet/Hosting#1421]
Given various issues and discussions around the design, as well as confusions about the lifetime and purposes of the individual interfaces (#1419, #1403, #1401, #1221), in my opinion a lot of parts should be revised and possibly re-designed for the next major release of the generic hosting infrastructure.

With the current design, I see the following problems:

### *Start* / *Stop* on `IHost`

The current `IHost` contract offers a *Start* and *Stop* method. This gives away the impression that the host can be started and stopped as many times as desired (essentially restarted), as long at it hasn't been disposed. However, since the `IApplicationLifetime` (and it's cancellation hooks) is only resolved once, the host ends up in an inconsistent state (where e.g. hosted services are restarted multiple times, but the host lifetime is not).

#### Proposal

Either a) get rid of the *Stop* method entirely or b) allow the cancellation hooks of the `IApplicationLifetime` to be reset.

### `IHostLifetime`

The purpose of the `IHostLifetime` is not really obvious. On one side, it kinda looks like a *hosting environment* which encapsulates the current type of the process. So in that regard, it's obvious that there must be **one**, and only **one**. However, what the implementations mostly do is hook themselves up to some application events upon start, and what to do if the application receives any external stop events (SIGTERM, service stop etc.), which can basically be done by any service which gets the `IApplicationLifetime` contract injected. So, why aren't there multiple host lifetimes?

#### Proposal

Either a) get rid of the `IHostLifetime` contract entirely. Instead, let the `HostBuilder` create a single, shared instance of `ApplicationLifetime` by itself, and offer a method to configure the lifetime. (Possibly even hooking the callbacks up into a delegate factory of the `IServiceCollection`, so that they can be configured with the awareness of an existing `IServiceProvider`.)

    IHostBuilder ConfigureLifetime(Action<HostBuilderContext, IApplicationLifetime> callback)

... or b) get rid of them and instead introduce a `IHostInitializer`, as proposed by #1401.

### Consistency in service lifetimes (resolving `IHost`)

In most applications which make extensive use of inversion of control, the DI container is the 'root' and major 'manager' of service lifetimes. It's the first thing being created, and the last thing being disposed (for services managed this way). In the current implementation of the hosting infrastructure, things aren't quite as obvious. The `HostBuilder` acts as a wrapper around the factory service which creates the DI container, but isn't responsible for disposing it afterwards. This is up to the `IHost` (through implementing `IDisposable`), but the `IHost` itself is registered within the DI container and resolved from it.

#### Proposal

Don't register the `IHost` within the `IServiceCollection`. Create it manually via the `HostBuilder`, and pass along an instance of the `IServiceProvider` for further resolving host lifetimes or the like. Dispose the `IServiceProvider` afterwards when disposing the `IHost`. That way, it's obvious what implementation is responsible for managing what lifetime. This would also solve the issue with the weird casting of the `ApplicationLifetime` which breaks the entire thing anyway when the implementation is being replaced (described in #169).

**Comment by davidfowl** [comment]
Some of these are great suggestions but they don’t affect the usage in any way. It merely cleans some implementation details.

The first one is breaking and we’ll never do it so better to make it work. The others are quality of code improvements that seem low priority IMO.

How does this fix those related bugs you linked?

**Comment by maelgrove** [comment]
@davidfowl Yes, those points indeed address the implementation details of certain concepts and not issues per se, since it's about resolving incosistencies and provide clear APIs that give away their intentions.

> The first one is breaking and we’ll never do it so better to make it work.

I don't really understand this tbh. What do you mean by 'we'll never do it'. The `IHost` contract offers methods that are required to be called in a certain way in a certain state, and can cause the application to end up in an unclean state otherwise, which means that a user has to rely on external documentation in order to have things work as intended.

> How does this fix those related bugs you linked?

The linked issues are not bugs either, but users who expressed confusion about a certain thing on the API or requested a certain feature which isn't quite obvious why it isn't there in the first place. The proposals provided would address certain parts of those issues. Actual bugs woulds be for instance #169, where simply swapping out the 'IApplicationLifetime' breaks the entire expected startup procedure.

**Comment by davidfowl** [comment]
> I don't really understand this tbh. What do you mean by 'we'll never do it'. The IHost contract offers methods that are required to be called in a certain order in a certain state, and can cause the application to end up in an unclean state otherwise, which means that a user has to rely on external documentation or should be aware of the implementation in order to have things work as intended.

I mean, we could look at what it means to make Start and Stop work. Removing APIs really isn't an option.

> The linked issues are not bugs either, but users who expressed confusion about a certain thing on the API or requested a certain feature which isn't quite obvious why it isn't there in the first place. The proposals provided would address certain parts of those issues. Actual bugs woulds be for instance #169, where simply swapping out the 'IApplicationLifetime' breaks the entire expected startup procedure.

If I'm being pedantic:

> https://github.com/aspnet/Hosting/issues/1403

That isn't confusing, we didn't add that feature intentionally. 

> https://github.com/aspnet/Hosting/issues/1401

This is the same as https://github.com/aspnet/Hosting/issues/1221 and this particular problem should be solved with IHostedService. I wouldn't exactly call the problem important unless more than one person has the issue.

> https://github.com/aspnet/Hosting/issues/1221

This is the same as 1401

**Comment by maelgrove** [comment]
@davidfowl #1403 was indeed a mistake on my side, since I confused `IHostingStartup` with something that post-configures the application after it has been built.

I think my point still stands. The API as well as internal things are made in a way a developer needs to be aware of the individual implementations if they would want to e.g. swap out something, as expressed by the reasons I stated in the first comment. After all, the entire infrastructure is built around a IoC container abstraction, and the APIs give away a sense of 'you can replace anything you want', which is not the case. I've updated my original issues with some clarifications etc.

**Comment by davidfowl** [comment]
I don’t disagree with some of the assessment it just has a low priority is all.

**Comment by muratg** [comment]
Tentatively putting this in 3.0.0.  Based on the priorities, we may end up moving it out.

**Comment by lmcarreiro** [comment]
This design is a little confusing, I asked a question here about it:
https://stackoverflow.com/questions/52413002/whats-the-difference-between-these-ways-to-start-run-a-generic-host-in-asp-net
I can call Run/Start on the `HostBuilder` class as well as on the `Host` class. And these two options (`Start` vs `Run`), are both of them really necessary?
